### ISSUE TITLE
Understanding frame extraction and target frames

### ISSUE THREAD

**Issue by Lumoria** [kohya-ss/musubi-tuner#59]
I appreciate the in-depth documentation in the dataset config readme, but I'm still not fully getting it. Can you help me to understand how target_frames in the config.toml works? With 100 frame videos to train on, if I set frame_extraction to 'head' and set target_frames to [1, 25, 45, 73], what exactly am I sending to the model? What frames are being sent, does it see all the frames out of 100? If it doesn't, how do i change target_frames so I send every frame? Does the frame rate of the video matter? I've gotten some pretty decent results from the previous settings, but the model is seemingly not picking up on content throughout videos randomly. 

This issue is is especially true with videos with lower frame rate anime / cartoon videos. In the case of training on a a set of anime animation videos with 28 frames I set target_frames = [1, 9, 17, 21]. The generated results from this have choppy frame pacing issues that don't match the input animation videos. 

Also when declaring datasets, I just separated them by total frame count. Should I also be further separating these by frame rate as well?

In summary I just want to ensure that I'm sending every frame from the input video to the training network.

**Comment by Lumoria** [comment]
> https://github.com/kohya-ss/musubi-tuner/blob/main/dataset/dataset_config.md#frame_extraction-options

Thanks for the reply, but as I stated I've already gone through this page many times and still don't understand the intended method of use for frame extraction.

```
head, target_frames = [1, 13, 25] -> extract head frames:
xooooooooooooooooooooooooooooooooooooooo
xxxxxxxxxxxxxooooooooooooooooooooooooooo
xxxxxxxxxxxxxxxxxxxxxxxxxooooooooooooooo
```

If each line correlates to each value in array 'target_frames', then this example makes me believe we are leaving the last 15 frames out of the training data. Why not just set the array to one value of '40' for a video with 40 frames? If we are after accurate motion replication, we would want to make sure we extract *all* the frames, right? This is where I must be misunderstanding how frame extraction works.

**Comment by tavyra** [comment]
Don't use head then, use slide with a stride, use head with the full length or use uniform to extract every other frame. If I had all 40-frame clips, I'd be using head [40] if they are all known equal. You can set multiple lengths with head if you want to generate videos not exactly 40 frames but always from the first, cleanest starting frame of the video (assuming you cut them correctly)

It's because it doesn't use the clip at all if you set 40 but some videos only have 38. You can define multiple datasets from the same .jsonl with various settings

**Comment by Lumoria** [comment]
> Don't use head then, use slide with a stride, use head with the full length or use uniform to extract every other frame. If I had all 40-frame clips, I'd be using head [40] if they are all known equal. You can set multiple lengths with head if you want to generate videos not exactly 40 frames but always from the first, cleanest starting frame of the video (assuming you cut them correctly)
> 
> It's because it doesn't use the clip at all if you set 40 but some videos only have 38. You can define multiple datasets from the same .jsonl with various settings

Ok cool, thanks for clearing up the head [40] case. I did actually test this with a single video of 100 frames and target_frames = [100] in which case the training samples weren't changing at all. But I noticed after the fact that it was probably because the video was omitted because 100 isn't a correct interval, right? I should have set target_frames = [97] to have a 100 frame clip be included right? Same thing with head and [40] where it would have to be [37] to include a 40 frame video and extract the first 37 frames?

**Comment by Lumoria** [comment]
> Frame values are rounded down to 4n+1. So 100 is same as 97, and 40 is same as 37.
> 
> If you run cache_latents.py with the argument `--debug_mode image`, you can check the first frame and the last frame from the frames to be extracted.

Ah, that is perfect, --debug_mode image will definitely help. Thanks for pointing this out!

**Comment by Sarania** [comment]
Also to better try to explain the various extraction methods with as simple language as possible. Firstly consider every N in [N, N, N] such as [1, 33, 65] as a pass that will be made through each of your input videos in the dataset. Each of them is a frame length that you want extracted from the source into a latent - a length bucket.

Note: In these examples I'm counting from "1" as the first frame of the video.

For the simplest "head" extraction which just extracts starting at the beginning, it will extract a 1 frame latent, a 33 frame latent, and a 65 frame latent - all starting at the beginning(frame 1), from each of your inputs. 

 "chunk" is also pretty simple. Consider for [13, 25] it will go through each of your inputs from the beginning extracting first as many 13 frame chunks as possible without overlap, then back to the beginning again taking as many 25 frame chunks without overlapping those. This is why 1 is excluded because that would take a single frame latent for every single frame. 

"slide" requires frame_stride be set as the spacing and considering for [1, 13] with a stride of 10, it will extract the first, 11th, 21st, 31st, etc for the 1 bucket pass. For the 13 bucket, it will extract the first 13 frames starting at the beginning, then it will extract 13 frames beginning at frame 11 inclusive, then it will extract 13 frames at frame 21 inclusive, etc... Overlap is allowed and will happen where N is greater than frame_stride.

"uniform" is the last and the most useful after "head", IMO. It requires frame_sample to be set as the number of samples to take for each length bucket. For [1, 13] with a frame_sample of 4, it will first take 4 equally spaced single frame latents from the input, and then 4 equally spaced 13 frame latents from the input. Overlap is allowed and will happen when N * frame_sample is greater than the length of the input.

Advanced note for "chunk" and "slide", if N (the length bucket we are processing) is greater than the number of frames left that haven't been processed yet, then it will simply move on rather than taking a too short sample or taking frames from the end of the last sample. For instance if we have a 33 frame input, and we are taking 13 frame chunks, we get two latents composed of frames 1-13, 14-26 and then frames 27-33 are ignored because there isn't enough for another chunk of 13. Similarly for slide, if there aren't at least frame_stride + N new frames after the end of the last sample, no new sample will be taken.

I hope this is helpful, it took me a while to wrap my head around it too but I think this is it - please feel free to correct me @kohya-ss

**Comment by FurkanGozukara** [comment]
@kohya-ss what do you recommend as target_frames

**Comment by shanque56666** [comment]
I am also confused about this dataset
Original Video, 40 frames: x = frame, o = no frame
oooooooooooooooooooooooooooooooooooooooo
Why are there only 30 letters "o", while the example states that there are 40 frames?

**Comment by Sarania** [comment]
"x" just means that frame will be sampled into a latent for training, "o" is frames that are not.

**Comment by ShanQueTit** [comment]
> "x" just means that frame will be sampled into a latent for training(along with any other "x" it's consecutive with/to), "o" is frames that are not.“x” 只是表示该帧将被采样到一个 latent 中进行训练（以及它与 “x” 连续/到的任何其他 “x”），“o” 是不是的帧。

thank

**Comment by rwfsmith** [comment]
If you want every frame, is that what frame_extraction = full does? At least, that's what I assume. I'm trying to train an anime lora with videos using that setting right now, so I hope it works as I expect.

**Comment by gambikules** [comment]
so how get each 2rd frame ? i understand nothing....

frame_extraction = "uniform"
[2,84]

**Comment by Straafe** [comment]
>4n+1

What is the technical reason behind this? I'm a situation with a 48GiB card where I am within 5GiB of being able to train, but having each sequence be a minimum of 5 frames is 1 frame too many, 4 would work, and I don't want to train on individual images.

**Comment by Sarania** [comment]
@Straafe A couple of things - you should more than be able to train any model Musubi supports on much less than 48GB(I have 16), though I don't know what your specific goals are. As far as your question, this has to do with how the video is represented in latent space. Hunyuan and Wan VAE both have a 4:1 compression ratio in the time domain - this means that every latent space frame will become 4 pixel space frames when decoded(and vice versa). This naturally limits the number of frames you can encode to those divisible by 4. The +1 comes from the model needing an "anchor frame" to start the sequence. So the actual possible values are like 1, 5, 9, 13, 17, 21, 25, 29, 33, 37 etc - it does actually increase in sequences of 4,  just with an offset of 1 from the anchor.

**Comment by Straafe** [comment]
@Sarania Thanks for this explanation, that's really informative. My dataset consists of 720p videos and I am training Wan 2.1  T2V 14B to start. My goal is not just trying to train in some sort of "style," (like you can do with just images in the dataset) but also the timing and _how_ things move/animate.  For that, I felt it is important to be able to keep a video dataset, but it is trying to use just over my available VRAM. I know I could lower the resolution of the dataset, but I am being greedy here because of the 48 GiB GPU.

**Comment by Sarania** [comment]
Are you using any of `--fp8_scaled`, `--blocks_to_swap`, `--compile` ? That would be the first place I'd start. fp8_scaled has no extra requirements and uses a scaled quantization to about half the models size with VERY little loss in quality, it's very good and highly recommended. Blocks to Swap also has no extra requirements and will keep some of the transformer blocks on CPU RAM and some on VRAM, allowing more space for big latents etc at a moderate but reasonable cost to speed. Compile can have some extra requirements like pytorch-triton and [on native Windows, a compiler](https://github.com/kohya-ss/musubi-tuner/pull/722#issuecomment-3535067203) but if you get it working it improves both speed and VRAM usage, potentially by a lot(WSL and Linux only need pytorch-triton). Also your optimizer choice can make a difference, dunno what you're using but came_pytorch is my personal choice for a good balance of VRAM/quality.

If you want to go even further than that there's my [blissful-tuner](https://github.com/Sarania/blissful-tuner/) which has `--rope_func comfy` which will replace the model's RoPE with one I ported from WanVideoWrapper that does not use complex numbers and so is faster and compiles better. It works best with compile but I've been told it improves things even without it and there's no major cost. My `--optimized_compile` also saves about 10-20% more VRAM than Musubi's native compile, but it can be tricky and doesn't work on Windows.  However, the combination of all these has allowed me to train Wan 2.1 at like [512, 512], [41] on 16GB VRAM + 64GB sysram on Linux! 

In reality I don't think it's necessary to go all the way to 720p based on my testing, but Wan does benefit from seeing higher res more than Hunyuan. If you're wanting to train also the movement/animation, then the longer length you can show the better. 5 frames is just too little IMO, 21 can start to be useful. I often do a combination of a low res but longer bucket like [256, 256], [81] and a higher res but shorter one like [512, 512], [41] so this is something you can consider too(Just list the same dataset twice in your toml datasets section, but specify a different cache dir and res/length for each one). Maybe include your 720p bucket at whatever amount of frames you can get with good settings, and then a bucket like 832x480 (Wan's 480p bucket size) with more/full length. That's what I would suggest to start, hopefully I'm not telling you things you already knew!

**Comment by Straafe** [comment]
Sorry for leaving that information out, yes I am currently using these arguments for optimization:

```
--fp8_base
--gradient_checkpointing
--gradient_checkpointing_cpu
--blocks_to_swap 20
```

I actually just added the last two arguments and it is fitting in the GPU now and successfully training, although extremely slowly (but not complaining!).

Thanks for the info and tips, I think that's a good idea to go with multiple datasets - some lower resolution with larger frame buckets and some high resolution single frames or small buckets. I'll check out the blissful tuner too. Right now I am basically using default everything. I'll let this run for a while and see what comes out the other end first, then probably move in the multiple dataset direction.

**Comment by Sarania** [comment]
I would /highly/ recommend adding `--fp8_scaled` in addition to `--fp8_base` - VRAM usage and speed will be very similar but overall quality will increase. With only fp8_base we just do a hard cast of the model to fp8_e4m3fn. With fp8_scaled, we do this more intelligently by calculating a scale factor that captures how the transformation in datatype affects the tensor, and we use this scale factor to reverse most of that change when we go to use the value. It's the difference between about 10% accuracy lost and about 2.5% in my measurements. For the user, the difference is a small time penalty up front to quantize the model (30-60 sec) and slightly more VRAM usage(because we keep a few params in full precision to maximize quality), but much higher quality. I personally start with the fp16 base model for Wan 2.1 and use fp8 scaled and this has given me phenomenal results! It might very slightly increase your VRAM usage, but you can compensate if need be with more blocks to swap (Wan max is 38 although over 36 is very slow in my experience) - ideally I try to land at about 95% of my VRAM (15GB of 16 in my case) to maximize speed while leaving a small buffer to avoid any possibility of mid training OOM.

Training at full res is gonna be SLOW. A normal Wan 2.1 training run of mine using like 512x512x41f and 480x272x81f with all my optimizations and flash attention 2 runs for ~2000 steps and takes about 8 hours. I have a 4070 Ti Super, fwiw. I do have a lot of experience with Wan 2.1 specifically (It's the model I've had the most luck training as far as video, Hunyuan has been okay and Wan 2.2 while giving me awesome results from the base I just can't get LoRA results I like from it) so if there's anything you wanna ask about I'm happy to share.

**Comment by Straafe** [comment]
I'm using bf16 as the base right now for Wan 2.1, should I be using fp16? I did set up my environment so --compile is working, and I added --fp8_scaled as well, and it is indeed very slow. I am prepared to let this run for days potentially though, so no problem there, for me it's more about getting it to run at all and with good enough data I can get the results I am after.
